Ical.Net/VTimeZoneProvider.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright ical.net project maintainers and contributors.
+// Licensed under the MIT license.
+
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using Ical.Net.CalendarComponents;
+using NodaTime;
+using NodaTime.TimeZones;
+
+namespace Ical.Net;
+
+/// <summary>
+/// Provides a custom <see cref="DateTimeZone"/> for a given TZID created from a <see cref="CalendarComponents.VTimeZone"/>
+/// </summary>
+public class VTimeZoneProvider : IDateTimeZoneProvider
+{
+    private readonly Dictionary<string, DateTimeZone> _customTimeZones;
+
+    /// <summary>
+    /// Creates a <see cref="VTimeZoneProvider"/> from a <see cref="Calendar"/>.
+    /// </summary>
+    /// <param name="calendar"></param>
+    /// <returns></returns>
+    public static VTimeZoneProvider FromCalendar(Calendar calendar) => FromTimeZone(calendar.TimeZones);
+
+    /// <summary>
+    /// Creates a <see cref="VTimeZoneProvider"/> from a collection of <see cref="VTimeZone"/>s.
+    /// </summary>
+    public static VTimeZoneProvider FromTimeZone(IEnumerable<VTimeZone> timeZones)
+    {
+        var provider = new VTimeZoneProvider();
+
+        foreach (var timeZone in timeZones)
+        {
+            if (TryCreateIntervals(timeZone, out var tzId, out var intervals))
+                provider.Add(tzId, intervals);
+        }
+
+        return provider;
+    }
+
+    private VTimeZoneProvider()
+    {
+        _customTimeZones = new();
+    }
+
+    private static bool TryCreateIntervals(VTimeZone vTimeZone, out string tzId, out List<ZoneInterval> intervals)
```

Comment:
When evaluating timezone intervals, RRULEs and RDATEs need to be considered. I suggest using `TimeZoneEvaluator` for that purpose.